### ISSUE TITLE
Initialization for bias parameters of the attention weights seems to be wrong

### ISSUE THREAD

**Issue by chrisway613** [fundamentalvision/Deformable-DETR#44]
Reference to the paper's decription:
  'Bias parameters of the linear projection are initialized to make A(mlqk) = 1/LK'

But in the code implementation:
https://github.com/fundamentalvision/Deformable-DETR/blob/11169a60c33333af00a4849f1808023eba96a931/models/ops/modules/ms_deform_attn.py#L72
we can see it was actually initialized to 0!

Due to the weight parameters are also initialized to 0, this will make the finally output becomes all zero..
Am I wrong.. so confused~

**Comment by lmc8133** [comment]
Have you ever run the code after your modify? What's the difference in the performance?

**Comment by Xuer0313** [comment]
[https://zhuanlan.zhihu.com/p/378418174](url) Hope it will help you

**Comment by Dstudying** [comment]
在知乎看到你的专栏，对于这个问题，我在代码里仔细看看了。实际上这里只是简单的初始化。    
https://github.com/fundamentalvision/Deformable-DETR/blob/11169a60c33333af00a4849f1808023eba96a931/models/ops/modules/ms_deform_attn.py#L71-L72
而在forward中，这里才使得A(mlqk) = 1/LK。 
https://github.com/fundamentalvision/Deformable-DETR/blob/11169a60c33333af00a4849f1808023eba96a931/models/ops/modules/ms_deform_attn.py#L99-L100

**Comment by xiaotaw** [comment]
GitHub Copilot解读

没有显式赋值为 1/(L·K)。这里用线性层生成 logits 后在最后一维做 softmax，使每个查询在所有 level×point 上归一化为总和1。

- 初始化时 attention_weights 的权重和偏置为0，logits全零 → softmax 输出均匀分布，即每项≈1/(L·K)。
- 训练过程中 logits 不再相等，softmax给出非均匀注意力，未强制保持 1/(L·K)。


> 而在forward中，这里才使得A(mlqk) = 1/LK。
> [Deformable-DETR/models/ops/modules/ms_deform_attn.py](https://github.com/fundamentalvision/Deformable-DETR/blob/11169a60c33333af00a4849f1808023eba96a931/models/ops/modules/ms_deform_attn.py#L99-L100)
> 
> Lines 99 to 100 in [11169a6](/fundamentalvision/Deformable-DETR/commit/11169a60c33333af00a4849f1808023eba96a931)
> 
>  attention_weights = self.attention_weights(query).view(N, Len_q, self.n_heads, self.n_levels * self.n_points) 
>  attention_weights = F.softmax(attention_weights, -1).view(N, Len_q, self.n_heads, self.n_levels, self.n_points)
